src/invert/laplace/impls/pcr_thomas/pcr_thomas.cxx

```diff
@@ -0,0 +1,1219 @@
+  }
+  int n = xe - xs + 1; // Number of X points on this processor,
+                       // including boundaries but not guard cells
+
+  a.reallocate(nmode, n);
+  b.reallocate(nmode, n);
+  c.reallocate(nmode, n);
+  xcmplx.reallocate(nmode, n);
+  bcmplx.reallocate(nmode, n);
+
+  xproc = localmesh->getXProcIndex(); // Local rank in x proc space
+  const int yproc = localmesh->getYProcIndex();
+  nprocs = localmesh->getNXPE();                    // Number of processors in x
+  myrank = yproc * nprocs + xproc;                  // Global rank for communication
+  n_mpi = localmesh->GlobalNxNoBoundaries / nprocs; // Number of internal x
+                                                    // grid points per proc
+}
+
+FieldPerp LaplacePCR_THOMAS::solve(const FieldPerp& rhs, const FieldPerp& x0) {
+  ASSERT1(localmesh == rhs.getMesh() && localmesh == x0.getMesh());
+  ASSERT1(rhs.getLocation() == location);
+  ASSERT1(x0.getLocation() == location);
+
+  FieldPerp x{emptyFrom(rhs)}; // Result
+
+  int jy = rhs.getIndex(); // Get the Y index
+  x.setIndex(jy);
+
+  // Get the width of the boundary
+
+  // If the flags to assign that only one guard cell should be used is set
+  int inbndry = localmesh->xstart;
+  int outbndry = localmesh->xstart;
+  if (((global_flags & INVERT_BOTH_BNDRY_ONE) != 0) || (localmesh->xstart < 2)) {
+    inbndry = outbndry = 1;
+  }
+  if ((inner_boundary_flags & INVERT_BNDRY_ONE) != 0) {
+    inbndry = 1;
+  }
+  if ((outer_boundary_flags & INVERT_BNDRY_ONE) != 0) {
+    outbndry = 1;
+  }
+
+  if (dst) {
+    BOUT_OMP(parallel) {
+      /// Create a local thread-scope working array
+      auto k1d = Array<dcomplex>(
+          localmesh->LocalNz); // ZFFT routine expects input of this length
+
+      // Loop over X indices, including boundaries but not guard cells. (unless periodic
+      // in x)
+      BOUT_OMP(for)
+      for (int ix = xs; ix <= xe; ix++) {
+        // Take DST in Z direction and put result in k1d
+
+        if (((ix < inbndry) && ((inner_boundary_flags & INVERT_SET) != 0)
+             && localmesh->firstX())
+            || ((localmesh->LocalNx - ix - 1 < outbndry)
+                && ((outer_boundary_flags & INVERT_SET) != 0) && localmesh->lastX())) {
+          // Use the values in x0 in the boundary
+          DST(x0[ix] + 1, localmesh->LocalNz - 2, std::begin(k1d));
+        } else {
+          DST(rhs[ix] + 1, localmesh->LocalNz - 2, std::begin(k1d));
+        }
+
+        // Copy into array, transposing so kz is first index
+        for (int kz = 0; kz < nmode; kz++) {
+          bcmplx(kz, ix - xs) = k1d[kz];
+        }
+      }
+
+      // Get elements of the tridiagonal matrix
+      // including boundary conditions
+      BOUT_OMP(for nowait)
+      for (int kz = 0; kz < nmode; kz++) {
+        BoutReal zlen = coords->dz * (localmesh->LocalNz - 3);
+        BoutReal kwave =
+            kz * 2.0 * PI / (2. * zlen); // wave number is 1/[rad]; DST has extra 2.
```

Comment:
Similar to what I asked for the `pcr` - why is there a 3 but the comment says 2?

Besides that, would that also work if `dz` is a `Field2D`?

I assume variation of `dz` along `z` wouldn't work due to FFTs require being uniform ...